File: lib/config/rush_yaml.dart
import 'dart:io';

import 'package:checked_yaml/checked_yaml.dart';
import 'package:json_annotation/json_annotation.dart';

part 'android.dart';
part 'kotlin.dart';

part 'rush_yaml.g.dart';

@JsonSerializable(
  anyMap: true,
  checked: true,
  disallowUnrecognizedKeys: true,
)
class RushYaml {
  @JsonKey(required: true)
  final String version;

  @JsonKey(name: 'dependencies')
  final List<String> runtimeDeps;

  @JsonKey(name: 'comptime_dependencies')
  final List<String> comptimeDeps;
  
  final String homepage;
  
  final String license;
  
  final bool desugar;
  
  final List<String> assets;
  
  final List<String> authors;
  
  final Android? android;
  
  final Kotlin? kotlin;

  RushYaml({
    required this.version,
    this.homepage = '',
    this.license = '',
    this.desugar = false,
    this.assets = const [],
    this.authors = const [],
    this.runtimeDeps = const [],
    this.comptimeDeps = const [],
    this.android,
    this.kotlin,
  });

  // ignore: strict_raw_type
  factory RushYaml._fromJson(Map json) => _$RushYamlFromJson(json);

  static Future<RushYaml> load(File configFile) async {
    try {
      return checkedYamlDecode(
          await configFile.readAsString(), (json) => RushYaml._fromJson(json!));
    } catch (e) {
      rethrow;
    }
  }
}

File: processor/processor/src/main/kotlin/io/shreyash/rush/processor/model/RushYaml.kt
package io.shreyash.rush.processor.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RushYaml(
    val version: String,
    val license: String = "",
    val homepage: String = "",
    val desugar: Boolean = false,
    val assets: List<String> = listOf(),
    val authors: List<String> = listOf(),
    @SerialName("dependencies") val runtimeDeps: List<String> = listOf(),
    @SerialName("comptime_dependencies") val compileDeps: List<String> = listOf(),
    val android: Android = Android(),
    val kotlin: Kotlin = Kotlin(),
)

@Serializable
data class Android(
    @SerialName("compile_sdk") val compileSdk: Int = 31,
    @SerialName("min_sdk") val minSdk: Int = 7,
)

@Serializable
data class Kotlin(
    @SerialName("compiler_version") val compilerVersion: String = "1.7.10",
)
